src/core/csv/reader.cc

```diff
@@ -419,46 +429,43 @@ py::oobj GenericReader::read_all(py::robj pysources)
   }
 
   if (!output_) {
-    throw IOError() << "Unable to read input " << src_arg.to_string();
+    throw IOError() << "Unable to read input " << *source_name;
   }
 
   job->done();
   return output_;
 }
 
 
-py::oobj GenericReader::read_buffer(const Buffer& buf, size_t extra_byte)
-{
-  if (logger) {
-    logger.invoke("debug", py::ostring("[1] Prepare for reading"));
-  }
-  open_buffer(buf, extra_byte);
-  job = std::make_shared<dt::progress::work>(WORK_PREPARE + WORK_READ);
-  bool done = read_jay();
-
-  if (!done) {
-    detect_and_skip_bom();
-    skip_to_line_number();
-    skip_to_line_with_string();
-    skip_initial_whitespace();
-    skip_trailing_whitespace();
-    job->add_done_amount(WORK_PREPARE);
-
-    read_empty_input() ||
-    detect_improper_files() ||
-    read_csv();
-  }
-
-  if (!output_) {
-    throw IOError() << "Unable to read input " << src_arg.to_string();
+void GenericReader::log_file_sample() {
+  if (!verbose) return;
+  trace("==== file sample ====");
+  const char* ch = sof;
+  bool newline = true;
+  for (int i = 0; i < 5 && ch < eof; i++) {
+    if (newline) trace("%s", repr_source(ch, 100));
+    else         trace("...%s", repr_source(ch, 97));
+    const char* start = ch;
+    const char* end = std::min(eof, ch + 10000);
+    while (ch < end) {
+      char c = *ch++;
+      // simplified newline sequence. TODO: replace with `skip_eol()`
+      if (c == '\n' || c == '\r') {
+        if ((*ch == '\r' || *ch == '\n') && *ch != c) ch++;
```

Comment:
Since we do `*ch++` above, shouldn't we also add a check `ch < end` here?

Reply:
good catch!